README
This is privbind, written by Shachar Shemesh with improvements by Amos
Shapira and Nadav Har'El.
Privbind is free software, released under the GNU General Public License
(see the attached file COPYING for details).

The project is currently managed by Jiri Horky (jiri.horky at gmail.com).


What is Privbind
================
Privbind is a tool for running a command as an unprivileged user, with
additional reserved port binding privileges.

Normally in Linux, only a superuser process can bind an Internet domain socket
with a privileged port (port numbers less than 1024). Accordingly, server
processes are typically run with superuser privileges, which can be dropped
after binding the reserved port. The privilege dropping has to be a part of the
server's code.

Privbind can run a server, right from the start, under an unprivileged non-root
user with just one extra privilege: the server can bind to reserved ports.

Privbind is useful in several situations. It can be used when the server code
cannot be trusted enough to start it as root; It can be used when the server
is written in a language without the setuid(2) feature (e.g., Java); It can
also be used to run a server written to remain running as root, while all the
privileges it really needs is the ability to bind reserved ports.


System requirements
===================
The current privbind implementation (see manual page privbind.1 for details)
was only tested on Linux, using a 2.6 kernel.

Among other things, it assumes the LD_PRELOAD feature, and the SOCK_SEQPACKET
socket type which is a relatively recent addition to Linux

How privbind works
==================
In theory, all one would need is to make use of Linux's capabilities(7) feature
and its CAP_NET_BIND_SERVICE. Unfortunately, Linux's capabilities feature is
totally useless because capability inheritance was never implemented!

Privbind uses a 100% user space approach for solving this issue. Privbind uses
the LD_PRELOAD method to override the running command's "bind" function.
Whenever the command tries to bind a privileged port, privbind intercepts the
call and forwards it to a process left behind, still running as root. This
process performs the bind on the command's behalf.

Forwarding of a socket to another process is done via a unix domain socket,
which supports sending a fd to another process.

privbind deviates from the accepted standard of "fork + exec", in that it is
the parent process, not the child, that does "exec". This is done so that the
command's return code (as well as actual return) be visible to the user
running privbind, and not privbind's return code, which is, for all intent and
purposes, irrelevant.

Caveats and corner cases
========================
Since privbind changes both the parent-child relationship and the code flow of
certain system functions, care must be taken to avoid the changes in semantics
from affecting the program run.

The most notable case for such interaction are signals. Two signals of note
should be pointed out. One is the case where privbind receives the "-n" option.
With this option, privbind's privileged process terminates after 'num'
privileged binds have been performed. This means that the child process
terminates while the parent is still running, causing the parent to receive a
SIGCHILD from a child it did not know it had. Privbind solves this problem by
performing a double-fork, and thus running with no parent.

The second case is when, as before, "-n" caused privbind to terminate prior to
the actual program terminating. If the program now attempts another bind, it is
possible that a SIGPIPE will be sent to the process. This signal, again, is one
not resulting from anything the program did, and may lead to trouble. The
solution is to send the message with the "MSG_NOSIGNAL" option, which prevents
SIGPIPE from being sent.

The third case of unintended consequences is when using privbind to run a
daemon. The standard instructions for writing daemons state that a daemon
should close, before starting to run, all of its open file descriptors.
Unfortunately, this also includes the fd used to communicate between privileged
and unprivileged processes. Privbind solves this problem by also intercepting
the "close" syscall, and making sure that our own fd is never closed.

For more information
====================

Refer to the manual page, privbind(1).
The latest version of this software can be found in
	https://github.com/JiriHorky/privbind

The old project webpage stays in
	http://sourceforge.net/projects/privbind


NEWS
21 January, 2015
Version 1.3 released
 - contains support for IPv6
 - enables keeping supplementary groups

13 June, 2007
First official release

With all known issues resolved, we are proud to announce version 1.01 of
privbind as ready for general use. Support is provided through the project's
mailing list at http://sourceforge.net/projects/privbind.


test.c
/*
 * privbind - allow unpriviledged apps to bind to priviledged ports
 * Copyright (C) 2007 Nadav Har'el
 * Copyright (C) 2007 Shachar Shemesh
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <netinet/in.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pwd.h>
#include <grp.h>

#include <pthread.h>

#define TEST_PORT 987

struct cmdoptions {
    int port; /* Port number to bind */
    int numbinds; /* Number of binds to perform */
    int wait; /* Number of seconds to wait after last bind */
    int processes; /* Number of processes to perform the bind from */
    int threads; /* Number of threads to perform the bind from */
} options={ TEST_PORT, 1, 0, 1, 1 };

void help( const char *progname )
{
    printf("testbin - Try to bind a privileged port\n");
    printf("Usage: %s [-n NUM] [-p NUM] [-w NUM]\n", progname);
    printf("\n");
    printf("-p NUM - Port number to bind. Default is %d\n", options.port);
    printf("-n NUM - Number of times to attempt bind. Default is %d\n", options.numbinds);
    printf("-w NUM - Number of seconds to wait after last bind.\n");
    printf("-P NUM - Number of processes to bind from.\n");
    printf("-T NUM - Number of threads to bind from.\n");
}

int parse_cmdline( int argc, char *argv[] )
{
    /* Fill in default values */
    int opt;

    while( (opt=getopt(argc, argv, "+n:w:p:hP:T:" ))!=-1 ) {
        switch(opt) {
        case 'n':
            options.numbinds=atoi(optarg);
            break;
        case 'w':
            options.wait=atoi(optarg);
            break;
        case 'p':
            options.port=atoi(optarg);
            break;
        case 'h':
            help(argv[0]);
            exit(0);
        case '?':
            help(argv[0]);
            exit(1);
        case 'P':
            options.processes=atoi(optarg);
            break;
        case 'T':
            options.threads=atoi(optarg);
            break;
        }
    }

    return optind;
}
int perform_bind( )
{
    int fd;
    struct sockaddr_in addr;

    fd = socket(PF_INET, SOCK_STREAM, 0);
    if (fd < 0) {
        perror("socket");
        return 0;
    }

    memset(&addr, 0, sizeof(addr));

    addr.sin_family = AF_INET;
    addr.sin_port = htons(options.port);
    addr.sin_addr.s_addr = htonl(INADDR_ANY);

    printf("Process %d thread %lx: About to bind\n", getpid(), pthread_self() );
    struct timeval tv1, tv2;
    gettimeofday(&tv1, NULL);
    int res=bind(fd, (struct sockaddr *)&addr, sizeof(addr));
    gettimeofday(&tv2, NULL);
    tv2.tv_sec-=tv1.tv_sec;
    if(tv1.tv_usec>tv2.tv_usec) {
       tv2.tv_sec--;
       tv2.tv_usec+=1000000;
    }
    tv2.tv_usec-=tv1.tv_usec;

    printf("Process %d thread %lx: Bind returned %d. Took %ld.%06ld secs\n", getpid(),
      pthread_self(), res, tv2.tv_sec, tv2.tv_usec );
    close(fd);

    if(res==-1) {
        perror("bind");
        return 0;
    }

    return 1;
}

void *perform_binds( void *p )
{
   int i=options.numbinds;

   printf("Process %d thread %lx performs %d binds\n", getpid(), pthread_self(), i);

   while( i-->0 ) {
      perform_bind(NULL);
   }

   return NULL;
}

int main(int argc, char *argv[])
{
    parse_cmdline( argc, argv );

    printf("running as user: %d (%s).\n", getuid(), getpwuid(getuid())->pw_name);
    printf("running as group: %d (%s).\n", getgid(), getgrgid(getgid())->gr_name);
    printf("running as process %d\n", getpid() );

    fflush(stdout);

    /* Make sure we have the correct concurrency level */
    int i;
    int parent=1;
    for( i=1; parent && i<options.processes; ++i ) {
       if( fork()==0 ) {
          printf("Created new process %d\n", getpid() );
          parent=0; /* We are not the ones who need to wait() at the end */
       }
    }

    pthread_t *threads=calloc(sizeof(pthread_t), options.threads-1);

    for( i=1; i<options.threads; ++i ) {
       pthread_create( threads+i-1, NULL, perform_binds, NULL );
    }
    /* And once more for the main thread */
    perform_binds(NULL);

    /* Sleep to check if the parent sticks around after the bind (if -n 1
       is used, it shouldn't */
    if( options.wait>0 ) {
        printf("Sleeping for %d seconds...\n", options.wait);
        sleep(options.wait);
    }

    /* Collect all started threads */
    for( i=1; i<options.threads; ++i ) {
       pthread_join(threads[i-1], NULL );

       printf("Process %d joined thread %lx\n", getpid(), threads[i-1]);
    }

    printf("Process %d done\n", getpid() );

    fflush(stdout);

    if( parent ) {
       for( i=1; i<options.processes; ++i ) {
          int status;

          int pid=wait(&status);

          printf("Reaped process %d\n", pid);
       }
    }

    return 0;
}


stub.h
/*
 * privbind - allow unpriviledged apps to bind to priviledged ports
 * Copyright (C) 2006 Shachar Shemesh
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#ifndef STUB_H
#define STUB_H

#include <dlfcn.h>

#define FUNCREDIR0( name, ret ) \
static ret stub_##name( void );\
static ret (* _##name)( void )=stub_##name;\
static ret stub_##name( void ) { _##name=dlsym( RTLD_NEXT, #name ); return _##name(); }

#define FUNCREDIR1( name, ret, type1 ) \
static ret stub_##name( type1 arg1 );\
static ret (* _##name)( type1 )=stub_##name;\
static ret stub_##name( type1 arg1 ) { _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1 ); }

#define FUNCREDIR2( name, ret, type1, type2 ) \
static ret stub_##name( type1 arg1, type2 arg2 );\
static ret (* _##name)( type1, type2 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2 ) { _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2 ); }

#define FUNCREDIR3( name, ret, type1, type2, type3 ) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3 );\
static ret (* _##name)( type1, type2, type3 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3 ) { _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3 ); }

#define FUNCREDIR4( name, ret, type1, type2, type3, type4 ) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4 );\
static ret (* _##name)( type1, type2, type3, type4 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4 )\
{ _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3, arg4 ); }

#define FUNCREDIR5( name, ret, type1, type2, type3, type4, type5 ) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5 );\
static ret (* _##name)( type1, type2, type3, type4, type5 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5 )\
{ _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3, arg4, arg5 ); }

#define FUNCREDIR6( name, ret, type1, type2, type3, type4, type5, type6 ) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6 );\
static ret (* _##name)( type1, type2, type3, type4, type5, type6 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6 )\
{ _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3, arg4, arg5, arg6 ); }

#define FUNCREDIR9( name, ret, type1, type2, type3, type4, type5, type6, type7, type8, type9 ) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6, type7 arg7, type8 arg8,\
	type9 arg9 );\
static ret (* _##name)( type1, type2, type3, type4, type5, type6, type7, type8, type9 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6, type7 arg7, type8 arg8,\
	type9 arg9 )\
{ _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9 ); }

#define FUNCREDIR10( name, ret, type1, type2, type3, type4, type5, type6, type7, type8, type9, type10) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6, type7 arg7, type8 arg8,\
	type9 arg9, type10 arg10);\
static ret (* _##name)( type1, type2, type3, type4, type5, type6, type7, type8, type9, type10)=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6, type7 arg7, type8 arg8,\
	type9 arg9, type10 arg10)\
{ _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10); }

#define FUNCREDIR12( name, ret, type1, type2, type3, type4, type5, type6, type7, type8, type9, type10, type11, type12 ) \
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6, type7 arg7, type8 arg8,\
	type9 arg9, type10 arg10, type11 arg11, type12 arg12 );\
static ret (* _##name)( type1, type2, type3, type4, type5, type6, type7, type8, type9, type10, type11, type12 )=stub_##name;\
static ret stub_##name( type1 arg1, type2 arg2, type3 arg3, type4 arg4, type5 arg5, type6 arg6, type7 arg7, type8 arg8,\
	type9 arg9, type10 arg10, type11 arg11, type12 arg12 )\
{ _##name=dlsym( RTLD_NEXT, #name ); return _##name( arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10, arg11, arg12 ); }

#endif // STUB_H


libprivbind.c
/*
 * privbind - allow unpriviledged apps to bind to priviledged ports
 * Copyright (C) 2006 Shachar Shemesh
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#include "config.h"

#define _GNU_SOURCE
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <pthread.h>

#include <stdlib.h>
#include <stdio.h>

#include "stub.h"

#include "ipc.h"

static int master_quit=0; /* Whether helper process quit - assume false at first */

FUNCREDIR1( close, int, int );
FUNCREDIR3( bind, int, int, const struct sockaddr *, socklen_t );

static void master_cleanup()
{
    /* The helper process has quit */
    master_quit=1;
    unsetenv("LD_PRELOAD");
    close(COMM_SOCKET);
}

/* Acquire or release the lock on the communication socket */
static int acquire_lock( int acquire )
{
   static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
   struct flock lock;

   lock.l_whence=SEEK_SET;
   lock.l_start=0;
   lock.l_len=0;

   if( acquire ) {
      if( pthread_mutex_lock(&mutex)!=0 )
      {
         /* The mutex is invalid. Assume we terminated */
         master_cleanup();

         return 0;
      }

      /* Acquire the fcntl lock on the socket. Need to retry in case of EINTR */
      lock.l_type=F_WRLCK;

      int res;
      while( (res=fcntl(COMM_SOCKET, F_SETLKW, &lock))!=0 && errno==EINTR )
         ;

      if( res!=0 )
         pthread_mutex_unlock(&mutex);

      return res==0;
   } else {
      /* Need to unlock, opposite order than locking */

      /* Release the fcntl lock */
      lock.l_type=F_UNLCK;
      int res;

      while( (res=fcntl(COMM_SOCKET, F_SETLKW, &lock))!=0 && errno==EINTR )
         ;

      pthread_mutex_unlock(&mutex);

      return res==0;
   }
}

int bind( int sockfd, const struct sockaddr *my_addr, socklen_t addrlen)
{
   /* First of all, attempt the bind. We only need the socket if it fails with access denied */

   int oret=_bind( sockfd, my_addr, addrlen );
   if( oret==0 || errno!=EACCES )
      return oret;

   /* In most cases, we can let the bind go through as is */
   if( master_quit || (my_addr->sa_family!=AF_INET && my_addr->sa_family!=AF_INET6) 
       || (my_addr->sa_family==AF_INET && addrlen<sizeof(struct sockaddr_in)) 
       || (my_addr->sa_family==AF_INET6 && addrlen<sizeof(struct sockaddr_in6))  
     ) { 
      errno=EACCES;
      return oret;
   }

   /* Prepare the ancillary data for passing the actual FD */
   struct msghdr msghdr={.msg_name=NULL};
   struct cmsghdr *cmsg;
   char buf[CMSG_SPACE(sizeof(int))];
   int *fdptr;

    msghdr.msg_control=buf;
    msghdr.msg_controllen=sizeof(buf);

    cmsg=CMSG_FIRSTHDR(&msghdr);
    cmsg->cmsg_level=SOL_SOCKET;
    cmsg->cmsg_type=SCM_RIGHTS;
    cmsg->cmsg_len=CMSG_LEN(sizeof(int));

    /* Initialize the payload */
    fdptr=(int *)CMSG_DATA(cmsg);
    fdptr[0]=sockfd;
    msghdr.msg_controllen=cmsg->cmsg_len;

    /* Don't forget the data component */
    struct ipc_msg_req request;
    struct iovec iov;

    msghdr.msg_iov=&iov;
    msghdr.msg_iovlen=1;

    iov.iov_base=&request;
    iov.iov_len=sizeof(request);

    request.type=MSG_REQ_BIND;

    /* Check family of the request, only INET and INET6 should make it here */
    if (my_addr->sa_family==AF_INET)
      request.data.bind4.addr=*(struct sockaddr_in *) my_addr;
    else if (my_addr->sa_family==AF_INET6)
      request.data.bind6.addr=*(struct sockaddr_in6 *) my_addr;

    int retval=oret;

    if( acquire_lock(1) ) {
       if( sendmsg( COMM_SOCKET, &msghdr, MSG_NOSIGNAL )>0 ) {
          /* Request was sent - wait for reply */
          struct ipc_msg_reply reply;

          if( recv( COMM_SOCKET, &reply, sizeof(reply), 0 )>0 ) {
             retval=reply.data.stat.retval;
             if( retval<0 )
                errno=reply.data.stat.error;
          } else {
             /* It would appear that the other process has closed, just return the original retval */
             master_cleanup();
          }
       } else {
          /* An error has occured! */
          if( errno==EPIPE || errno==ENOTCONN || errno==EBADF ) {
             master_cleanup();
          } else {
             perror("privbind communication socket error");
             master_cleanup();
          }
       }

       acquire_lock(0);
    }

    /* Make sure we return the original errno, regardless of what caused us to fail */
    if( retval!=0 )
        errno=EACCES;

    return retval;
}

int close(int fd)
{
   /* Is it our fd? */
   if( fd!=COMM_SOCKET )
      /* No - pass it on to "close" */
      return _close(fd);

   /* Yes - override the close and return an error */
   errno=EBADF;
   return -1;
}


main.c
/*
 * privbind - allow unpriviledged apps to bind to priviledged ports
 * Copyright (C) 2006-2007 Shachar Shemesh
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>

#include <errno.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "config.h"
#include "ipc.h"

#define FALSE (0!=0)
#define TRUE (0==0)
#define MAX_GROUPS 32
#define UNDEF (-1)

struct cmdoptions {
    uid_t uid; /* UID to turn into */
    gid_t gid; /* GID to turn into */
    gid_t groups[MAX_GROUPS]; /* Supplemenetary groups to set up */
    int   num_groups; /* number of supplementary gruops */
    char disable_supp_groups; /*Whether to set supplementary groups or not */
    int numbinds; /* Number of binds to catch before program can make do on its own */
    const char *libname; /* Path to library to use as preload */
#if DEBUG_TESTING
    int wait; /* Time to artificially prolong the bind time by */
#endif
} options;

void usage( const char *progname )
{
    fprintf(stderr, "Usage: %s -u UID [-g GID] [-G] [-n NUM] command [arguments ...]\n", progname);
    fprintf(stderr, "Run '%s -h' for more information.\n", progname);
}
void help( const char *progname )
{
    printf("%s - run a program as an unpriviledged user, while still being\n",
	   PACKAGE_STRING);
    printf("     able to bind to low ports.\n");
    printf("Usage: %s -u UID [-g GID] [-n NUM] command [arguments ...]\n", progname);
    printf("\n"
	"-u - Name or id of user to run as (mandatory)\n"
	"-g - Name or id of group to run as (default: the user's default group)\n"
	"-G - Disables setting user's supplementary groups (default: enabled)\n"
	"-n - number of binds to catch. After this many binds have happened,\n"
        "     the helper proccess exits.\n"
        "-l - Explicitly specify the path to the library to use for preload\n"
        "     This option is for debug use only.\n"
#if DEBUG_TESTING
        "-w - Delay each bind by num seconds. Only useful for internal privbind\n"
        "     testing.\n"
#endif
	"-h - This help screen\n");
}

int parse_cmdline( int argc, char *argv[] )
{
    /* Fill in default values */
    options.numbinds=0;
    options.uid=0;
    options.gid=-1;
    memset(options.groups, UNDEF, MAX_GROUPS);
    options.num_groups = MAX_GROUPS;
    options.disable_supp_groups=0;
    options.libname=PKGLIBDIR "/" PRELOADLIBNAME;
    char * username;
    
    int opt;

    while( (opt=getopt(argc, argv, "+n:u:g:Gl:w:h" ))!=-1 ) {
        switch(opt) {
        case 'n':
            options.numbinds=atoi(optarg);
	    if (options.numbinds < 0) {
	        fprintf(stderr, "Illegal number of binds passed: '%s'\n",
		  optarg);
		exit(1);
	    }
            break;
        case 'u':
            {
                username = optarg;
                struct passwd *pw=getpwnam(username);
                if( pw!=NULL ) {
                    options.uid=pw->pw_uid;
                } else {
                    options.uid=atoi(optarg);
                    char * errp = 1;
		    options.uid= strtol(username,  &errp, 10);
                    if (errp == username || errp != NULL) {
                        fprintf(stderr, "Could not resolve username %s as string nor as UID\n", username);
                        exit(1);
                    }
                    pw = getpwuid(options.uid);
                    if( pw == NULL ) {
                        fprintf(stderr, "User with UID '%d' not found\n", options.uid);
                        exit(1);
                    } else {
                        username = pw->pw_name;
                    }
                }
                /* set the user's default group */
                if( options.gid==(gid_t)UNDEF ) {
                    options.gid=pw->pw_gid;
                }
                if ( ! options.disable_supp_groups ) {
                    if ( getgrouplist(pw->pw_name, pw->pw_gid, options.groups, &(options.num_groups)) == -1) {
                        fprintf(stderr, "Error getting list of groups for user %s. Maybe he is member of too too many groups? The limit is %d.\n", optarg, MAX_GROUPS); 
                        exit(1);
                    }
                }
            }
            break;
        case 'g':
            {
                if ( *optarg == '\0' ) {
                    fprintf(stderr, "Empty group parameters\n");
                    exit(1);
                }
                struct group *gr=getgrnam(optarg);
                if( gr!=NULL ) {
                    options.gid=gr->gr_gid;
                } else {
                    char *endptr;
                    options.gid=strtol(optarg, &endptr, 10);
                    if( *endptr != '\0') {
                        fprintf(stderr, "Group name '%s' not found\n", optarg);
                        exit(1);
                    }
                    if( options.gid==(gid_t)UNDEF ) {
                        fprintf(stderr, "Illegal group id %d\n", options.gid);
                        exit(1);
                    }
                }
            }
            break;
        case 'G':
            options.disable_supp_groups=1;
            memset(options.groups, UNDEF, MAX_GROUPS);
            break;
        case 'l':
            options.libname=optarg;
            break;
#if DEBUG_TESTING
        case 'w':
            options.wait=atoi(optarg);
            break;
#endif
        case 'h':
            help(argv[0]);
            exit(0);
        case '?':
            usage(argv[0]);
            exit(1);
        }
    }

    if(options.uid==0){
        fprintf(stderr, "Missing UID (-u) option.\n");
        usage(argv[0]);
        exit(1);
    }
    if(options.gid==(gid_t)UNDEF){
        fprintf(stderr, "Unable to find group which to run as.\n");
        usage(argv[0]);
        exit(1);
    }

    if( (argc-optind)<=0 ) {
        fprintf(stderr, "ERROR: missing a command to run.\n");
        usage(argv[0]);
        exit(1);
    }
    return optind;
}

/* Technically speaking, the "child" is the parent process. Internally, we call it by its semantics
 * rather than by its function.
 */
int process_child( int sv[2], int argc, char *argv[] )
{
    /* Drop privileges */
    if(options.disable_supp_groups) {
            if( setgroups(0, NULL )<0 ) {
                perror("privbind: error dropping supplementary groups");
                return 2;
            }
    } else {
        if (setgroups(options.num_groups, options.groups)  != 0) {
                perror("privbind: error setting supplementary groups");
                return 2;
        }
    }

    if( setgid(options.gid)<0 ) {
        perror("privbind: setgid");
        close(sv[0]);
        return 2;
    }
    if( setuid(options.uid)<0 ) {
        perror("privbind: setuid");
        close(sv[0]);
        return 2;
    }

    /* Close the parent socket */
    close(sv[1]);

    /* Rename the child socket to the pre-determined fd */
    if( dup2(sv[0], COMM_SOCKET)<0 ) {
        perror("privbind: dup2");
        return 2;
    }
    close(sv[0]);

    /* Set the LD_PRELOAD environment variable */
    char *ldpreload=getenv("LD_PRELOAD");
    if( ldpreload==NULL ) {
        setenv("LD_PRELOAD", options.libname, FALSE );
    } else {
        char *newpreload=malloc(strlen(ldpreload)+strlen(options.libname)+2); /* One extra for the ":", another for the NULL */
        if( newpreload==NULL ) {
            fprintf(stderr, "privbind: Error creating preload environment - out of memory\n");
            return 2;
        }

        sprintf( newpreload, "%s:%s", options.libname, ldpreload );

        setenv("LD_PRELOAD", newpreload, TRUE );

        free(newpreload);
    }

    /* Set up the variables for exec */
    char **new_argv=calloc(argc+1, sizeof(char*) );
    if( new_argv==NULL ) {
        fprintf(stderr, "privbind: Error creating new command line: out of memory\n");
        return 2;
    }

    int i;
    for( i=0; i<argc; ++i ) {
        new_argv[i]=argv[i];
    }

    execvp(new_argv[0], new_argv);
    perror("privbind: exec");
    return 2;
}

/* See comment for "process_child" regarding reverse roles */
int process_parent( int sv[2] )
{
    /* Some of the run options mean that we terminate before our "child". We don't want to confuse
     * the child with SIGCHLD of which it is not aware.
     */
    int grandchild_pid=fork();

    if( grandchild_pid==-1 ) {
        perror("privbind: Error creating grandchild process");

        return 1;
    }

    if( grandchild_pid!=0 ) {
        /* We are the grandchild's parent. Terminate cleanly to indicate to our parent that it's ok
         * to start the actual program.
         */
        return 0;
    }

    /* Close the child socket */
    close(sv[0]);

    /* Don't hold on to resources that we will not use */
    chdir("/");
    close(0);
    close(1);
    /* do not close(2) as we still use stderr */
    
    /* Don't be killed by signals sent to the previous process group */
    setsid();

    /* wait for request from the child */
    do {
        struct msghdr msghdr={.msg_name=NULL};
        struct cmsghdr *cmsg;
        char buf[CMSG_SPACE(sizeof(int))];
        struct ipc_msg_req request;
        struct iovec iov;
        struct ipc_msg_reply reply = {.type=MSG_REP_NONE};
        int recvbytes;

        msghdr.msg_control=buf;
        msghdr.msg_controllen=sizeof(buf);

        iov.iov_base = &request;
        iov.iov_len = sizeof request;

        msghdr.msg_iov = &iov;
        msghdr.msg_iovlen = 1;

        if ( (recvbytes = recvmsg( sv[1], &msghdr, 0)) > 0) {
            if ((cmsg = (struct cmsghdr *)CMSG_FIRSTHDR(&msghdr)) != NULL) {
                switch (request.type) {
                case MSG_REQ_NONE:
                    reply.type = MSG_REP_NONE;
                    if (send(sv[1], &reply, sizeof reply, 0) != sizeof reply)
                        perror("privbind: send");
                    break;
                case MSG_REQ_BIND:
                    reply.type = MSG_REP_STAT;
                    int sock;
                    if (cmsg->cmsg_len == CMSG_LEN(sizeof(int))
                            && cmsg->cmsg_level == SOL_SOCKET
                            && cmsg->cmsg_type == SCM_RIGHTS)
                        sock = *((int*)CMSG_DATA(cmsg));
                    else {
                        sock = -1;
                    }
                    int size;
                    if ( ((struct sockaddr)request.data.bind.addr).sa_family == AF_INET ) {
                        size = sizeof(struct sockaddr_in);
                    } else if ( ((struct sockaddr)request.data.bind.addr).sa_family == AF_INET6) {
                        size = sizeof(struct sockaddr_in6);
                    } else {
                        perror("privbind: Invalid inet family type");
                        break;
                    }

                    reply.data.stat.retval =
                        bind(sock, (struct sockaddr *)&request.data.bind.addr, size);
                    if (reply.data.stat.retval < 0)
                        reply.data.stat.error = errno;
                    
#if DEBUG_TESTING
                    /* Sleep to check for races */
                    if( options.wait!=0 )
                        sleep(options.wait);
#endif

                    if (send(sv[1], &reply, sizeof reply, 0) != sizeof reply)
                        perror("privbind: send");
                    if (sock > -1 && close(sock))
                        perror("privbind: close");
                    break;
                default:
                    fprintf(stderr, "privbind: bad request type: %d\n",
                            request.type);
                    break;
                }
            } else {
                fprintf(stderr, "privbind: empty request\n");
            }
        } else if (recvbytes == 0) {
            /* If the child closed its end of the socket, it means the
               child has exited. We have nothing more to do. */

            return 0;
        } else {
            perror("privbind: recvmsg");
        }
    } while (options.numbinds == 0 || --options.numbinds > 0);


    /* If we got here, the child has done the number of binds
       specified by the -n option, and we have nothing more to do
       and should exit, leaving behind no helper process */

    return 0;
}

int main( int argc, char *argv[] )
{
    int skipcount=parse_cmdline( argc, argv );
    int ret=0;

    /* Warn if we're run as SUID */
    if( getuid()!=geteuid() ) {
        fprintf(stderr, "!!!!Running privbind SUID is a security risk!!!!\n");
    }

    /* Create a couple of sockets for communication with our children */
    int sv[2];
    if( socketpair(AF_UNIX, SOCK_SEQPACKET, 0, sv)<0 ) {
        perror("privbind: socketpair");
        return 2;
    }

    pid_t child_pid=fork();
    /* http://sourceforge.net/mailarchive/message.php?msg_name=20070601194744.GA29875%40fermat.math.technion.ac.il
     * Reverse the usual role of "parent" and "child".
     * Parent process perform "child" actions - running the command.
     * Child process is the one that listens on the socket and handles the binds
     */
    switch(child_pid) {
    case -1:
        perror("privbind: fork");
        exit(1);

    case 0:
        /* We are the child */
        ret=process_parent( sv );
        break;
    default:
        /* We are the parent */

        {
            /* Wait for the child to exit. */
            int status=0;

            do {
                waitpid( child_pid, &status, 0 );
            } while( !WIFEXITED(status) && !WIFSIGNALED(status) );

            if( WIFEXITED(status) ) {
                ret=WEXITSTATUS(status);

                if( ret==0 ) {
                    /* Child has indicated that it is ready */
                    ret=process_child( sv, argc-skipcount, argv+skipcount );
                }
            } else {
                fprintf(stderr, "privbind: root process terminated with signal %d\n", WTERMSIG(status) );
                ret=2;
            }
        }
        break;
    }

    return ret;
}
